File: typescript/src/agent.ts
import "dotenv/config";
import * as readline from "node:readline/promises";

import {
  AIMessage,
  HumanMessage,
  ToolMessage,
  type BaseMessage,
} from "@langchain/core/messages";

import { buildAgent } from "./graph.js";

function preview(text: string, max: number): string {
  return text.length > max ? text.slice(0, max) + "…" : text;
}

function printNewMessages(messages: BaseMessage[], startIndex: number): void {
  for (const msg of messages.slice(startIndex)) {
    if (msg instanceof AIMessage) {
      const toolCalls = msg.tool_calls ?? [];
      for (const tc of toolCalls) {
        const args = preview(JSON.stringify(tc.args ?? {}), 120);
        console.log(`  [tool] ${tc.name}(${args})`);
      }
      if (typeof msg.content === "string" && msg.content.length > 0) {
        console.log(`agent > ${msg.content}`);
      }
    } else if (msg instanceof ToolMessage) {
      console.log(`  [tool result] ${preview(String(msg.content), 240)}`);
    }
  }
}

async function main(): Promise<void> {
  console.log("Loading agent…");
  const { agent } = await buildAgent();
  console.log();
  console.log("Subconscious × Natoma agent. Type 'exit' or Ctrl-D to quit.\n");

  const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout,
  });

  let messages: BaseMessage[] = [];
  try {
    while (true) {
      const userInput = (await rl.question("you > ")).trim();
      if (!userInput) continue;
      if (["exit", "quit"].includes(userInput.toLowerCase())) break;

      messages.push(new HumanMessage(userInput));
      const start = messages.length;
      const result = await agent.invoke({ messages });
      messages = result.messages as BaseMessage[];
      printNewMessages(messages, start);
      console.log();
    }
  } finally {
    rl.close();
  }
}

main().catch((err) => {
  console.error(err);
  process.exit(1);
});

File: python/agent.py
"""Interactive REPL for the Natoma × Subconscious starter agent."""
import asyncio
import json

from dotenv import load_dotenv
from langchain_core.messages import AIMessage, HumanMessage, ToolMessage

from src.graph import build_agent


def _print_new_messages(messages, start_index: int) -> None:
    for msg in messages[start_index:]:
        if isinstance(msg, AIMessage):
            for tc in msg.tool_calls or []:
                args = json.dumps(tc.get("args", {}), default=str)
                if len(args) > 120:
                    args = args[:120] + "…"
                print(f"  [tool] {tc['name']}({args})")
            if msg.content:
                print(f"agent > {msg.content}")
        elif isinstance(msg, ToolMessage):
            output = str(msg.content)
            if len(output) > 240:
                output = output[:240] + "…"
            print(f"  [tool result] {output}")


async def chat_loop(agent) -> None:
    print("Subconscious × Natoma agent. Type 'exit' or Ctrl-D to quit.\n")
    messages: list = []
    while True:
        try:
            user_input = input("you > ").strip()
        except (EOFError, KeyboardInterrupt):
            print()
            return
        if not user_input:
            continue
        if user_input.lower() in {"exit", "quit"}:
            return

        messages.append(HumanMessage(content=user_input))
        start = len(messages)
        result = await agent.ainvoke({"messages": messages})
        messages = result["messages"]
        _print_new_messages(messages, start)
        print()


async def main() -> None:
    load_dotenv()
    print("Loading agent…")
    agent, _mcp_client = await build_agent()
    print()
    await chat_loop(agent)


if __name__ == "__main__":
    asyncio.run(main())
